// iso-crypto/dist/iso/lib/types.js
export var Ciphers = /*#__PURE__*/ function(Ciphers) {
    Ciphers["AES"] = "AES";
    return Ciphers;
}({});
export var Modes = /*#__PURE__*/ function(Modes) {
    Modes["CBC"] = "CBC";
    Modes["CTR"] = "CTR";
    return Modes;
}({});
export var Sizes = /*#__PURE__*/ function(Sizes) {
    Sizes[Sizes["KEY_128"] = 128] = "KEY_128";
    Sizes[Sizes["KEY_160"] = 160] = "KEY_160";
    Sizes[Sizes["KEY_192"] = 192] = "KEY_192";
    Sizes[Sizes["KEY_256"] = 256] = "KEY_256";
    Sizes[Sizes["KEY_384"] = 384] = "KEY_384";
    Sizes[Sizes["KEY_512"] = 512] = "KEY_512";
    return Sizes;
}({});
export const defaultEncryption = {
    cipher: "AES",
    size: 256,
    mode: "CTR"
};
export var Algorithms = /*#__PURE__*/ function(Algorithms) {
    Algorithms["SHA1"] = "SHA1";
    Algorithms["SHA2"] = "SHA2";
    Algorithms["RAW"] = "raw";
    return Algorithms;
}({});
export const defaultHash = {
    algorithm: "SHA2",
    size: 256
};
export var Encodings = /*#__PURE__*/ function(Encodings) {
    Encodings["BASE64"] = "base64";
    Encodings["BASE64URL"] = "base64url";
    Encodings["HEX"] = "hex";
    Encodings["RAW"] = "raw";
    Encodings["UTF8"] = "utf8";
    return Encodings;
}({});
export const defaultEncoding = "utf8";
export var Curves = /*#__PURE__*/ function(Curves) {
    Curves["P256"] = "p256";
    Curves["P384"] = "p384";
    Curves["P521"] = "p521";
    return Curves;
}({});
export const defaultCurve = "p256";
